### PR TITLE
Implement `Vec::pop_if`

### PR DESCRIPTION
**Pull request by igneousflock**
This PR adds `Vec::pop_if` to the public API, behind the `vec_pop_if` feature.

```rust
impl<T> Vec<T> {
    pub fn pop_if<F>(&mut self, f: F) -> Option<T>
        where F: FnOnce(&mut T) -> bool;
}
```

Tracking issue: #122741

## Open questions

- [ ] Should the first unit test be split up?
- [ ] I don't see any guidance on ordering of methods in impl blocks, should I move the method elsewhere?

### CODE REVIEW

File: library/alloc/src/vec/mod.rs
Comment (cynecx) [inline library/alloc/src/vec/mod.rs:2089]: This needs some type of drop guard because `f` might panic and that would've run `T`'s `drop` without adjusting the length of the vec.

Comment (avandesa) [reply library/alloc/src/vec/mod.rs:2089]: Makes sense, is there an example of that pattern I could reference?

Comment (cynecx) [reply library/alloc/src/vec/mod.rs:2089]: You can probably just wrap the last value in a `ManuallyDrop`, that should do it too.

**Review by joboet — CHANGES_REQUESTED** [review]
If the predicate modifies the value but returns `false`, that should be reflected in the vector. Hence this should call `f` with the value still in the vector, and only then remove it. Also, I'd prefer if this didn't use `unsafe`, I'd wager that the optimizer is able to deal with a combination of `slice::last` and `Vec::pop` just fine.

**Comment by avandesa** [comment]
Fair enough, I'll trust your judgement on the optimizer. Switch to `last_mut` and `pop` and added a test for an impure predicate.

@rustbot review

**Review by joboet — APPROVED** [review]
Looks great, thank you very much! Could you squash the commits, please?

Then r=me 
(this shorthand tells other people with merge rights that they can merge this on my behalf if don't get to it right away)

**Comment by avandesa** [comment]
Ope, I misunderstood that I think

**Pull request merged**
